tests/ttnn/unit_tests/operations/test_moreh_matmul.py

```diff
@@ -0,0 +1,256 @@
+    passing = moreh_matmul(params, False, None, device)
+    assert passing
+
+
+@pytest.mark.parametrize(
+    "params",
+    (
+        # input, weight, bias(1d or scalar), output
+        ([32, 32], [32, 32], [32, 32], False, False),  # single-core
+        (
+            [2, 3, 1, 2, 3, 2, 64, 64],
+            [2, 1, 4, 2, 1, 2, 64, 64],
+            [2, 3, 4, 2, 3, 2, 64, 64],
+            False,
+            False,
+        ),  # batched matmul
+    ),
+)
+def test_moreh_matmul_enable_cache(params, device, use_program_cache):
+    device.enable_program_cache()
+    torch.manual_seed(3072)
+    for i in range(4):
+        # change input's transpose option
+        if i % 2 == 1:
+            param_list = list(params)
+            param_list[3] = False if param_list[3] else True
+            params = tuple(param_list)
+        passing = moreh_matmul(params, False, None, device)
+        assert passing
+    assert device.num_program_cache_entries() == 2
+    device.disable_and_clear_program_cache()
+
+
+@pytest.mark.parametrize(
+    "params",
+    (
+        # input, other, output shape, transpose input, other
+        ([32, 3200], [3200, 32], [32, 32], False, False),
+        ([3100, 31], [3100, 31], [31, 31], True, False),
+    ),
+)
+@pytest.mark.parametrize("compute_kernel_options", compute_kernel_options, ids=compute_kernel_ids)
+def test_moreh_matmul_fp32_dest_acc(params, compute_kernel_options, device):
+    torch.manual_seed(3072)
+    input_shape, other_shape, output_shape, transpose_input, transpose_other = params
+    tt_input, tt_other, tt_output, _, _, _, torch_input, torch_other, _ = get_tensors(
+        input_shape, other_shape, output_shape, False, False, False, device, use_randint=False
+    )
+
+    compute_kernel_config = get_compute_kernel_options(compute_kernel_options)
+
+    torch_input = torch_input.transpose(-1, -2) if transpose_input else torch_input
+    torch_other = torch_other.transpose(-1, -2) if transpose_other else torch_other
+
+    # tt matmul
+    cpu_layout = ttnn.ROW_MAJOR_LAYOUT
+    tt_output = ttnn.operations.moreh.matmul(
+        tt_input,
+        tt_other,
+        transpose_input=transpose_input,
+        transpose_other=transpose_other,
+        output=tt_output,
+        compute_kernel_config=compute_kernel_config,
+    )
+    tt_output_cpu = tt_output.cpu().to(cpu_layout).unpad_from_tile(output_shape).to_torch()
+
+    # torch matmul (float)
+    torch_out = torch.matmul(torch_input.float(), torch_other.float())
+
+    # test for equivalance
+    rtol = atol = 0.1
+    passing, output_pcc = comp_allclose_and_pcc(torch_out, tt_output_cpu, pcc=0.999, rtol=rtol, atol=atol)
+    logger.debug(f"Out passing={passing}")
+    logger.debug(f"Output pcc={output_pcc}")
+    diff = torch.abs(torch_out - tt_output_cpu)
+    logger.debug(f"std={torch.std(diff)}")
+    logger.debug(f"mean={diff.mean()}")
+    logger.debug(f"topk(5) {torch.topk(diff.reshape(-1), 5)}")
+
+    # TODO
+    # assert passing
```

Comment:
I’ve left some comments regarding the `preserve_fp32_precision` flag in `moreh_matmul_program_factory.cpp`. This is related to the `fp32 dest mode` and this test method.

Currently, this test method passes unconditionally, but it needs to be updated. (refer to issue #9410 for more details).
Without setting this flag, in the second case of this test method, the MAE between the CPU and TT is even worse in `fp32 mode` than in bf16 mode, compared to the CPU reference.

Could you please modify the test code as follows?
1. In one test case, calculate the results with `fp32 mode` off and on (that is, remove the compute_kernel_options parameter and call `ttnn.operations.moreh.matmul` twice within the method, toggling the` fp32 mode`), and compare the MAE to check if there’s an improvement.
2. Compare the output tensor from `fp32 mode` on with the reference tensor using `comp_allclose_and_pcc`, and adjust the pcc threshold from 0.999 to 0.99.

The test case should pass only if both conditions (1 and 2) are true.
If any part needs more explanation, please let me know.

Reply:
I am testing, fp32 is passing but bf16 is currently failing.

Reply:
Please check if the new tests are what you expect, @dongjin-na

Reply:
Sorry for late reply.
Yes, bf16 has failed, and fp32 has passed. Could you adjust the bf16 case to reflect that failure is the correct outcome?”

Reply:
Is `assert diff.mean() < diff_fp16.mean()` good enough, or should I explicitly put something like `assert not passing`?